shared/manifest.go

```diff
@@ -68,27 +71,95 @@ func (m *Manifest) unmarshalAll() error {
 	return nil
 }
 
-// Contains checks whether m contains the path.
-func (m *Manifest) Contains(path string) (bool, error) {
-	if m.imap == nil {
-		if err := m.unmarshalAll(); err != nil {
-			return false, err
-		}
+func trieContains(t interface{}, parts []string) interface{} {
+	if len(parts) == 0 {
+		return t
 	}
-	parts := strings.Split(path, "/")
-	// Split always returns at least one element.
-	// Remove the leading empty part.
-	if parts[0] == "" {
-		parts = parts[1:]
+
+	// t could be nil (e.g. if the previous part does not exist in the map), in which case casting will fail.
+	trie, ok := t.(map[string]interface{})
+	if !ok {
+		return nil
+	}
+	return trieContains(trie[parts[0]], parts[1:])
+}
+
+// ContainsFile checks whether m contains a file path (including directories).
+func (m *Manifest) ContainsFile(path string) (bool, error) {
+	if err := m.unmarshalAll(); err != nil {
+		return false, err
 	}
+
+	path = strings.Trim(path, "/")
+	if path == "" {
+		// Root directory always exists.
+		return true, nil
+	}
+	parts := strings.Split(path, "/")
 	for _, items := range m.imap {
-		if trieContains(items, parts) {
+		if trieContains(items, parts) != nil {
 			return true, nil
 		}
 	}
 	return false, nil
 }
 
+// ContainsTest checks whether m contains a full test URL.
+func (m *Manifest) ContainsTest(testURL string) (bool, error) {
+	if err := m.unmarshalAll(); err != nil {
```

Comment:
This is a good question. The reasons are twofold:

1. I could use the same logic as filtering, but the major difference between them is the use case. Filtering is usually only done once (e.g. in an API request to get a subtree), but existence tests tend to be done multiple times in a row (e.g. in wpt-metadata to check whether a bunch of paths exist). If I use the exact same logic, then we're likely to parse the same subtrees multiple times. Parsing the whole trie recursively is an easy optimization based on use cases.
2. In theory I could implement a true JIT cache so that each subtree is parsed at most once, but that'd require a big type switch (to check whether each `interface{}` is a parsed map or a `RawMessage`). And I think at that point it'd be premature optimization.